File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/additives/adapter/AdditivesAdapter.java
package openfoodfacts.github.scrachx.openfood.features.additives.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.models.entities.additive.AdditiveName;

/**
 * Created by prajwalm on 16/04/18.
 */
public class AdditivesAdapter extends RecyclerView.Adapter<AdditiveViewHolder> {
    private final List<AdditiveName> additives;
    private final ClickListener clickListener;

    public AdditivesAdapter(List<AdditiveName> additives, ClickListener clickListener) {

        this.additives = additives;
        this.clickListener = clickListener;
    }

    public interface ClickListener {
        void onClick(int position, String name);
    }

    @NonNull
    @Override
    public AdditiveViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.additives_item, parent, false);
        return new AdditiveViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull AdditiveViewHolder holder, int position) {

        holder.additiveName.setText(additives.get(position).getName());
        holder.itemView.setOnClickListener(view -> {
            if (clickListener != null) {
                int pos = holder.getAdapterPosition();
                clickListener.onClick(pos, additives.get(pos).getName());
            }
        });
    }

    @Override
    public int getItemCount() {
        return additives.size();
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/additives/AdditivesAdapter.kt
package openfoodfacts.github.scrachx.openfood.features.additives

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import openfoodfacts.github.scrachx.openfood.databinding.AdditivesItemBinding
import openfoodfacts.github.scrachx.openfood.models.entities.additive.AdditiveName
import openfoodfacts.github.scrachx.openfood.utils.AutoUpdatableAdapter

/**
 * Created by prajwalm on 16/04/18.
 */
class AdditivesAdapter(
    initialValue: List<AdditiveName> = emptyList(),
    private val clickListener: ((String) -> Unit)?,
) : RecyclerView.Adapter<AdditivesAdapter.ViewHolder>(), AutoUpdatableAdapter {

    var additives: List<AdditiveName> by autoNotifying(
        initialValue = initialValue,
        comparator = compareBy { it.id }
    )

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = AdditivesItemBinding.inflate(inflater, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val name = additives[position].name

        holder.additiveName.text = name
        holder.itemView.setOnClickListener {
            clickListener?.invoke(name)
        }
    }

    override fun getItemCount() = additives.count()

    class ViewHolder(binding: AdditivesItemBinding) : RecyclerView.ViewHolder(binding.root) {
        val additiveName = binding.additiveName
    }
}

